### ISSUE TITLE
Compilation fails at csharp_generator.cc

### ISSUE THREAD

**Issue by vschiavoni** [grpc/grpc#4505]
From current head   compilation hangs here:

```
src/compiler/csharp_generator.cc:47:43: error: 'google::protobuf::compiler::csharp::GetUmbrellaClassName' has not been declared
 using google::protobuf::compiler::csharp::GetUmbrellaClassName;
                                           ^
src/compiler/csharp_generator.cc: In function 'void grpc_csharp_generator::{anonymous}::GenerateServiceDescriptorProperty(grpc::protobuf::io::Printer*, const ServiceDescriptor*)':
src/compiler/csharp_generator.cc:237:62: error: 'GetUmbrellaClassName' was not declared in this scope
              "umbrella", GetUmbrellaClassName(service->file()), "index",
                                                              ^
make: *** [/grpc/objs/opt/src/compiler/csharp_generator.o] Error 1
```

Installation:
```
gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.8/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.8.4-2ubuntu1~14.04' --with-bugurl=file:///usr/share/doc/gcc-4.8/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.8 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.8 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libmudflap --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.8-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04)
```

**Comment by vschiavoni** [comment]
Can it something specific to ubuntu ? The compilation goes smoothly under the latest ubuntu (docker image ubuntu:latest).

**Comment by jtattermusch** [comment]
This is most likely because you are using a wrong version of protobuf.   The GetUmbrellaClassName was introduced I believe in protobuf 3.0.0-beta1 release.  It looks like you might have an older version of protobuf installed   and the compilation picks up the wrong public protobuf headers.

It seems unlikely this would be an ubuntu specific issue.

**Comment by dbovis** [comment]
Looks like the commit: 
https://github.com/google/protobuf/commit/a6361a124a6537c69aaaa4e667372ba4e89c1fd1

has changed from GetUmbrellaClassName to GetReflectionClassName

Change the code in src/compiler/csharp_generator.cc to use the new class name and it will compile at least.

**Comment by Andyccs** [comment]
@jtattermusch You might be wrong because I get the same error message when using protobuf v3.0.0-beta-1.1

**Comment by Andyccs** [comment]
@dbovis suggestion works at least

**Comment by jtattermusch** [comment]
Looks like this is indeed a duplicate of issue https://github.com/grpc/grpc/issues/4213.

**Comment by jtattermusch** [comment]
Now that #4787 is merged, this should be fixed.

**Comment by emmenlau** [comment]
Release 0.12.0 still has this problem, does not compile under Ubuntu 15.04 unless @dbovis suggestion is used.

**Comment by jtattermusch** [comment]
@emmenlau, this will be fixed in the next release, the PR #4787 was not part of 0.12.

**Comment by emmenlau** [comment]
I see, ok thanks! I'll include the workaround in our build system, but it might be nice to make a quick followup-release on 0.12.0 so that other people can compile... :-)
